vpr/src/pack/verify_clustering.cpp

```diff
@@ -0,0 +1,444 @@
+                    // rectangles of the regions are equal.
+                    if (calc_cluster_region == cluster_region) {
+                        found_region = true;
+                        break;
+                    }
+                }
+                if (!found_region) {
+                    VTR_LOG_ERROR(
+                        "Cluster block %zu cluster constraint does not match "
+                        "the intersection of all constrained blocks in it\n",
+                        size_t(clb_blk_id));
+                    num_errors++;
+                    break;
+                }
+            }
+        }
+    }
+    return num_errors;
+}
+
+unsigned verify_clustering(const ClusteredNetlist& clb_nlist,
+                           const AtomNetlist& atom_nlist,
+                           const AtomLookup& atom_lookup,
+                           const vtr::vector<ClusterBlockId, std::unordered_set<AtomBlockId>>& clb_atoms,
+                           const vtr::vector<ClusterBlockId, PartitionRegion>& cluster_constraints,
+                           const UserPlaceConstraints& constraints) {
+    unsigned num_errors = 0;
+    // Check that every cluster has an entry in the clb_atoms vector.
+    if (clb_atoms.size() != clb_nlist.blocks().size()) {
+        VTR_LOG_ERROR(
+            "The number of clusters in the clb_atoms does not match the number "
+            "of clusters in the netlist.\n");
+        num_errors++;
+        // Return here since this error can cause serious issues below.
+        return num_errors;
+    }
+    // Check conssitency between which clusters the atom's think thet are in and
+    // which atoms the clusters think they have.
+    num_errors += check_clustering_atom_consistency(clb_nlist,
+                                                    atom_nlist,
+                                                    atom_lookup,
+                                                    clb_atoms);
+    // Check that the cluster pbs and atom pbs are consistent with each other.
+    // Each cluster should have a root pb, each atom should be a leaf descendent
+    // of its cluster's pb.
+    num_errors += check_clustering_pb_consistency(clb_nlist,
+                                                  atom_nlist,
+                                                  atom_lookup);
+    // Check that all clusters are clustering atoms which are constrained to
+    // overlapping regions and the constraints of the clusters are consistent
+    // with the atoms they contain.
+    num_errors += check_clustering_floorplanning_consistency(clb_nlist,
+                                                             clb_atoms,
+                                                             cluster_constraints,
+                                                             constraints);
+    // TODO: There exists more checks for the clustering in base/check_netlist.cpp
```

Comment:
I think general netlist features like mapping consistency are good to put in the netlist verify routine if possible. But I guess given that the atom_lookup clb mapping is not in the netlist or tied to it, it's OK to have the checks here instead.  Just suggesting you think about whether we should have a lookup verifier that can be called by this routine for checks that make sense ...